### ISSUE TITLE
Weird validation and train values after upgrade from 0.21.5 to 0.22

### ISSUE THREAD

**Issue by nfsecurity** [ml-explore/mlx-lm#18]
Hi, I have been doing fine tuning to Gemma 2 model with MLX for several months and every training process started like:

% mlx_lm.lora --train --model ./mlx_model -c config.yaml
Trainable parameters: 0.048% (4.473M/9241.706M)
Starting training..., iters: 400
Iter 1: Val loss 3.770, Val took 57.861s
Iter 5: Val loss 3.044, Val took 57.845s
Iter 5: Train loss 3.361, Learning Rate 1.000e-05, It/sec 0.165, Tokens/sec 282.648, Trained Tokens 8557, mem 56.494 GB
Iter 5: Saved adapter weights to adapters/adapters.safetensors and adapters/0000005_adapters.safetensors.
Iter 10: Val loss 2.526, Val took 58.144s
Iter 10: Train loss 2.669, Learning Rate 1.000e-05, It/sec 0.160, Tokens/sec 282.493, Trained Tokens 17373, mem 58.946 GB
Iter 10: Saved adapter weights to adapters/adapters.safetensors and adapters/0000010_adapters.safetensors.
Iter 15: Val loss 2.368, Val took 57.714s
Iter 15: Train loss 2.530, Learning Rate 1.000e-05, It/sec 0.048, Tokens/sec 98.130, Trained Tokens 27571, mem 65.006 GB
Iter 15: Saved adapter weights to adapters/adapters.safetensors and adapters/0000015_adapters.safetensors.
Iter 20: Val loss 2.094, Val took 57.665s
Iter 20: Train loss 2.135, Learning Rate 1.000e-05, It/sec 0.189, Tokens/sec 281.103, Trained Tokens 35004, mem 65.006 GB
Iter 20: Saved adapter weights to adapters/adapters.safetensors and adapters/0000020_adapters.safetensors.

But after the mlx-lm upgrade to 0.22 using the command pip3 install -U mlx-lm, validation and train loss values started slow down in a very different way:

% mlx_lm.lora --train --model ./mlx_model -c config.yaml
Starting training..., iters: 400
Iter 1: Val loss 3.170, Val took 66.146s
Iter 5: Val loss 2.388, Val took 62.995s
Iter 5: Train loss 2.871, Learning Rate 1.000e-05, It/sec 0.047, Tokens/sec 86.523, Trained Tokens 9278, mem 68.671 GB
Iter 5: Saved adapter weights to adapters/adapters.safetensors and adapters/0000005_adapters.safetensors.
Iter 10: Val loss 0.994, Val took 64.346s
Iter 10: Train loss 1.646, Learning Rate 1.000e-05, It/sec 0.148, Tokens/sec 287.029, Trained Tokens 18947, mem 68.671 GB
Iter 10: Saved adapter weights to adapters/adapters.safetensors and adapters/0000010_adapters.safetensors.
Iter 15: Val loss 0.203, Val took 65.001s
Iter 15: Train loss 0.354, Learning Rate 1.000e-05, It/sec 0.165, Tokens/sec 284.542, Trained Tokens 27563, mem 68.671 GB
Iter 15: Saved adapter weights to adapters/adapters.safetensors and adapters/0000015_adapters.safetensors.
Iter 20: Val loss 0.167, Val took 67.138s
Iter 20: Train loss 0.173, Learning Rate 1.000e-05, It/sec 0.162, Tokens/sec 286.342, Trained Tokens 36422, mem 68.671 GB
Iter 20: Saved adapter weights to adapters/adapters.safetensors and adapters/0000020_adapters.safetensors.

Train and validation loss have decreased very early. I did a pip install mlx-lm==0.21.5 again and I have the old behavior that is the behavior that I expected for training (validation and training loss decrease slow).

I did no changes in train, validation or test dataset (1k about, 60/20/20), also no changes in hyperparameters (rank 32, alpha 64, learning 1e-05, batch size 4, iters 400, num_layers -1), all data and configuration are the same, the only change is the mlx-lm version.

Is this a normal behavior in the new 0.22 update?, thank you!

**Comment by awni** [comment]
Could you share the full config and what kind of dataset you are using? Is it a completion/chat/text dataset?

Also good to double check the trainable parameters is the same for both (it should be printed at the very top):

> Trainable parameters: 0.048% (4.473M/9241.706M)

**Comment by nfsecurity** [comment]
Hi @awni, thank you for your reply.

I double checked the output, for 0.22.0 is:

% mlx_lm.lora --train --model ./mlx_model -c config.yaml
Loading configuration file config.yaml
Loading pretrained model
Loading datasets
Training
Trainable parameters: 0.048% (4.473M/9241.706M)
Starting training..., iters: 400
Iter 1: Val loss 3.174, Val took 57.537s
Iter 5: Val loss 2.363, Val took 58.254s
Iter 5: Train loss 2.787, Learning Rate 1.000e-05, It/sec 0.161, Tokens/sec 276.084, Trained Tokens 8557, Peak 56.494 GB
Iter 5: Saved adapter weights to adapters/adapters.safetensors and adapters/0000005_adapters.safetensors.
Iter 10: Val loss 1.009, Val took 58.104s
Iter 10: Train loss 1.547, Learning Rate 1.000e-05, It/sec 0.161, Tokens/sec 283.585, Trained Tokens 17373, Peak 59.975 GB
Iter 10: Saved adapter weights to adapters/adapters.safetensors and adapters/0000010_adapters.safetensors.
Iter 15: Val loss 0.211, Val took 57.675s
Iter 15: Train loss 0.386, Learning Rate 1.000e-05, It/sec 0.035, Tokens/sec 72.244, Trained Tokens 27571, Peak 64.930 GB
Iter 15: Saved adapter weights to adapters/adapters.safetensors and adapters/0000015_adapters.safetensors.
Iter 20: Val loss 0.167, Val took 57.454s
Iter 20: Train loss 0.204, Learning Rate 1.000e-05, It/sec 0.191, Tokens/sec 283.746, Trained Tokens 35004, Peak 64.930 GB
Iter 20: Saved adapter weights to adapters/adapters.safetensors and adapters/0000020_adapters.safetensors.

Then, I went back to 0.21.5:

% pip3 install mlx-lm==0.21.5
Defaulting to user installation because normal site-packages is not writeable
Collecting mlx-lm==0.21.5
  Using cached mlx_lm-0.21.5-py3-none-any.whl.metadata (9.9 kB)
Requirement already satisfied: mlx>=0.22.0 in ./Library/Python/3.9/lib/python/site-packages (from mlx-lm==0.21.5) (0.23.2)
Requirement already satisfied: numpy in ./Library/Python/3.9/lib/python/site-packages (from mlx-lm==0.21.5) (1.26.4)
Requirement already satisfied: transformers>=4.39.3 in ./Library/Python/3.9/lib/python/site-packages (from transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (4.49.0)
Requirement already satisfied: protobuf in ./Library/Python/3.9/lib/python/site-packages (from mlx-lm==0.21.5) (4.25.5)
Requirement already satisfied: pyyaml in ./Library/Python/3.9/lib/python/site-packages (from mlx-lm==0.21.5) (6.0.2)
Requirement already satisfied: jinja2 in ./Library/Python/3.9/lib/python/site-packages (from mlx-lm==0.21.5) (3.1.4)
Requirement already satisfied: filelock in ./Library/Python/3.9/lib/python/site-packages (from transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (3.16.1)
Requirement already satisfied: huggingface-hub<1.0,>=0.26.0 in ./Library/Python/3.9/lib/python/site-packages (from transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (0.26.3)
Requirement already satisfied: packaging>=20.0 in ./Library/Python/3.9/lib/python/site-packages (from transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (24.2)
Requirement already satisfied: regex!=2019.12.17 in ./Library/Python/3.9/lib/python/site-packages (from transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (2024.11.6)
Requirement already satisfied: requests in ./Library/Python/3.9/lib/python/site-packages (from transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (2.32.3)
Requirement already satisfied: tokenizers<0.22,>=0.21 in ./Library/Python/3.9/lib/python/site-packages (from transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (0.21.0)
Requirement already satisfied: safetensors>=0.4.1 in ./Library/Python/3.9/lib/python/site-packages (from transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (0.4.5)
Requirement already satisfied: tqdm>=4.27 in ./Library/Python/3.9/lib/python/site-packages (from transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (4.67.1)
Requirement already satisfied: sentencepiece!=0.1.92,>=0.1.91 in ./Library/Python/3.9/lib/python/site-packages (from transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (0.2.0)
Requirement already satisfied: MarkupSafe>=2.0 in ./Library/Python/3.9/lib/python/site-packages (from jinja2->mlx-lm==0.21.5) (2.1.5)
Requirement already satisfied: fsspec>=2023.5.0 in ./Library/Python/3.9/lib/python/site-packages (from huggingface-hub<1.0,>=0.26.0->transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (2024.9.0)
Requirement already satisfied: typing-extensions>=3.7.4.3 in ./Library/Python/3.9/lib/python/site-packages (from huggingface-hub<1.0,>=0.26.0->transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (4.12.2)
Requirement already satisfied: charset-normalizer<4,>=2 in ./Library/Python/3.9/lib/python/site-packages (from requests->transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (3.4.0)
Requirement already satisfied: idna<4,>=2.5 in ./Library/Python/3.9/lib/python/site-packages (from requests->transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (3.10)
Requirement already satisfied: urllib3<3,>=1.21.1 in ./Library/Python/3.9/lib/python/site-packages (from requests->transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (1.26.15)
Requirement already satisfied: certifi>=2017.4.17 in ./Library/Python/3.9/lib/python/site-packages (from requests->transformers>=4.39.3->transformers[sentencepiece]>=4.39.3->mlx-lm==0.21.5) (2024.8.30)
Using cached mlx_lm-0.21.5-py3-none-any.whl (152 kB)
Installing collected packages: mlx-lm
  Attempting uninstall: mlx-lm
    Found existing installation: mlx-lm 0.22.0
    Uninstalling mlx-lm-0.22.0:
      Successfully uninstalled mlx-lm-0.22.0
Successfully installed mlx-lm-0.21.5

And did the training again:

% mlx_lm.lora --train --model ./mlx_model -c config.yaml
Loading configuration file config.yaml
Loading pretrained model
Loading datasets
Training
Trainable parameters: 0.048% (4.473M/9241.706M)
Starting training..., iters: 400
Iter 1: Val loss 3.770, Val took 57.861s
Iter 5: Val loss 3.044, Val took 57.845s
Iter 5: Train loss 3.361, Learning Rate 1.000e-05, It/sec 0.165, Tokens/sec 282.648, Trained Tokens 8557, Peak 56.494 GB
Iter 5: Saved adapter weights to adapters/adapters.safetensors and adapters/0000005_adapters.safetensors.
Iter 10: Val loss 2.526, Val took 58.144s
Iter 10: Train loss 2.669, Learning Rate 1.000e-05, It/sec 0.160, Tokens/sec 282.493, Trained Tokens 17373, Peak58.946 GB
Iter 10: Saved adapter weights to adapters/adapters.safetensors and adapters/0000010_adapters.safetensors.
Iter 15: Val loss 2.368, Val took 57.714s
Iter 15: Train loss 2.530, Learning Rate 1.000e-05, It/sec 0.048, Tokens/sec 98.130, Trained Tokens 27571, Peak 65.006 GB
Iter 15: Saved adapter weights to adapters/adapters.safetensors and adapters/0000015_adapters.safetensors.
Iter 20: Val loss 2.094, Val took 57.665s
Iter 20: Train loss 2.135, Learning Rate 1.000e-05, It/sec 0.189, Tokens/sec 281.103, Trained Tokens 35004, Peak 65.006 GB
Iter 20: Saved adapter weights to adapters/adapters.safetensors and adapters/0000020_adapters.safetensors.

This is my config.yml file:

model: "./mlx_model"
train: true
fine_tune_type: lora
data: "dataset"
seed: 0
num_layers: -1
batch_size: 4
iters: 400
val_batches: 25
learning_rate: 1e-05
steps_per_report: 5
steps_per_eval: 5
resume_adapter_file: null
adapter_path: "adapters"
save_every: 5
test: false
test_batches: -1
max_seq_length: 2048
grad_checkpoint: false
lora_parameters:
  rank: 32
  alpha: 64
  scale: 10.0
  dropout: 0.0
lr_schedule: null

My dataset is a SPANISH custom created dataset of 1k internal corporate chats that contains unethical behavior, like fraud, corruption, abuse and compliance issues and the purpose of training is to format the output based on the fraud triangle theory, identifying pressure, opportunity and rationalization expressions, for example:

"Hola Jose, procede con mucha cautela, es mejor que no te vean los jefes sacando dinero de la caja menor, hazlo rapido y antes de que los auditores hagan el arqueo, no vaya a ser que se la pillen y nos hechen a todos, Ya le dije pues, no me haga quedar mal frente a todos" 

-------> 

"Sí, el mensaje es relevante para la identificación de riesgos de fraude interno. Existe un comportamiento fraudulento porque se esta mencionando sacar dinero de la caja menor sin que nadie lo vea. Las personas estan usando una frase en especial (procede con mucha cautela) cuando saben que algo no esta bien y tienen miedo de ser descubiertas. Segun la teoria del triangulo del fraude de Donald Cressey (presion, oportunidad y justificacion), esta persona aprovechará la oportunidad de sacar dinero de la caja menor haciéndolo rápido sin que sus jefes lo vean y antes de que los auditores hagan el arqueo"

Thank you again!

**Comment by awni** [comment]
Can you also tell me which model it is? Looks like you are using a custom conversion of a gemma 2? What size, what quantization etc?

**Comment by nfsecurity** [comment]
Hi @awni, it is Gemma 2 9B it directly from google at huggingface:

% mlx_lm.convert --hf-path google/gemma-2-9b-it
% mlx_lm.lora --train --model ./mlx_model -c config.yaml

**Comment by awni** [comment]
I'm getting a very different number for trainable parameters than you are with the same config and model. Could you please double check your setup and make sure you provided the right information:

```
Trainable parameters: 0.194% (17.891M/9241.706M)
```

The command I ran is:
```
mlx_lm.lora --config lora_config.yaml
```

The config is:

```yaml
model: "./mlx_model"
train: true
fine_tune_type: lora
data: "mlx-community/wikisql"
seed: 0
num_layers: -1
batch_size: 4
iters: 400
val_batches: 25
learning_rate: 1e-05
steps_per_report: 5
steps_per_eval: 5
resume_adapter_file: null
adapter_path: "adapters"
save_every: 5
test: false
test_batches: -1
max_seq_length: 2048
grad_checkpoint: false
lora_parameters:
  rank: 32
  alpha: 64
  scale: 10.0
  dropout: 0.0
lr_schedule: null
```

**Comment by nfsecurity** [comment]
Hi @awni, I made this video to reproduce the the weird behavior: https://www.nofraud.la/wp-content/uploads/2025/03/mlx-lm.mp4

I have been testing with different hyperparameters (trying to figure out why I am having this behavior). This time I am running the training with this config.yaml:

```
model: "./mlx_model"
train: true
fine_tune_type: lora
data: "dataset"
seed: 0
num_layers: -1
batch_size: 4
iters: 400
val_batches: 25
learning_rate: 1e-05
steps_per_report: 5
steps_per_eval: 5
resume_adapter_file: null
adapter_path: "adapters"
save_every: 5
test: false
test_batches: -1
max_seq_length: 2048
grad_checkpoint: false
lora_parameters:
  rank: 16
  alpha: 32
  scale: 10.0
  dropout: 0.0
lr_schedule: null
```

I am still getting very different results through iterations (train and validation loss) between 0.21.5 and 0.22.0 without changing anything (dataset is the same, config.yaml is the same).

Thank you some much for your time!

**Comment by awni** [comment]
There is something strange going on here and it seems to be related to boolean masking https://github.com/ml-explore/mlx-examples/pull/1319.

I will dig in further and let you know. Thanks for the test cases.

**Comment by nfsecurity** [comment]
Thank you.
